src/kvstore/comm.h

```diff
@@ -526,28 +526,50 @@ class CommDevice : public Comm {
     }
 
     InitBuffersAndComm(src);
+    auto& stage = stage_buf_[key];
     auto& buf = merge_buf_[key];
-    std::vector<NDArray> reduce(src.size());
-    CopyFromTo(src[0], &(buf.merged), priority);
-    reduce[0] = buf.merged;
 
-    if (buf.copy_buf.empty()) {
-      // TODO(mli) this results in large device memory usage for huge ndarray,
-      // such as the largest fullc in VGG. consider to do segment reduce with
-      // NDArray.Slice or gpu direct memory access. for the latter, we need to
-      // remove some ctx check, and also it reduces 20% perf
-      buf.copy_buf.resize(src.size()-1);
-      for (size_t i = 0; i < src.size()-1; ++i) {
-        buf.copy_buf[i] = NDArray(
-          buf.merged.shape(), buf.merged.ctx(), false, buf.merged.dtype());
-      }
+    if (buf.merged.is_none()&& stage.copy_buf.empty()) {
+        stage.copy_buf.resize(src.size()-1);
+        for (size_t i=0; i < src.size()-1; ++i)
+            stage.copy_buf[i] = NDArray(
+              stage.merged.shape(), stage.merged.ctx(), false, stage.merged.dtype());
     }
-    for (size_t i = 0; i < src.size()-1; ++i) {
-      CopyFromTo(src[i+1], &(buf.copy_buf[i]), priority);
-      reduce[i+1] = buf.copy_buf[i];
+    std::vector<NDArray> reduce_s(stage.copy_buf.size()+1);
+    for (size_t i = 0, j = 0; i < src.size(); ++i) {
+      int id = src[i].ctx().dev_id;
+      if ((!buf.merged.is_none() && id == stage.merged.ctx().dev_id)
+            || (buf.merged.is_none() && i == 0)) {
+        CopyFromTo(src[i], &(stage.merged), priority);
+        reduce_s[0] = stage.merged;
+      } else if (id >= 4 || buf.merged.is_none()) {
+        CopyFromTo(src[i], &(stage.copy_buf[j]), priority);
+        reduce_s[j+1] = stage.copy_buf[j];
+        j++;
+      }
     }
+    ElementwiseSum(reduce_s, &stage.merged);
+    // Main reduce result on gpu 0 including the partial result from gpu 4
+    if (!buf.merged.is_none()) {
+        std::vector<NDArray> reduce(buf.copy_buf.size()+1);
+        for (size_t i = 0, j = 0; i < src.size(); ++i) {
+          int id = src[i].ctx().dev_id;
+          if (id == buf.merged.ctx().dev_id) {
+            CopyFromTo(src[i], &(buf.merged), priority);
+            reduce[0] = buf.merged;
+          } else if (id < 4) {
+              CopyFromTo(src[i], &(buf.copy_buf[j]), priority);
+              reduce[j+1] = buf.copy_buf[j];
+              j++;
+          }
+        }
 
-    ElementwiseSum(reduce, &buf.merged);
+        CopyFromTo(stage.merged, &(buf.copy_buf[buf.copy_buf.size()-1]), priority);
+        reduce[reduce.size()-1] = buf.copy_buf[buf.copy_buf.size()-1];
+        ElementwiseSum(reduce, &buf.merged);
```

Comment:
Missing priority? Should this line be `ElementwiseSum(reduce, &buf.merged, priority);`?